File: app/src/main/java/beans/prj/newsapp/di/AppModule.java
package beans.prj.newsapp.di;

import android.content.Context;
import beans.prj.newsapp.NewsApplication;
import beans.prj.newsapp.data.AppRepository;
import beans.prj.newsapp.data.RetrofitClient;
import beans.prj.newsapp.ui.base.ViewModelProviderFactory;
import dagger.Module;
import dagger.Provides;
import retrofit2.Retrofit;

import javax.inject.Singleton;

@Module
public class AppModule {

    @Singleton
    @Provides
    Context provideContext(NewsApplication application) {
        return application;
    }

    @Singleton
    @Provides
    ViewModelProviderFactory provideViewModelProviderFactory(AppRepository repository) {
        return new ViewModelProviderFactory(repository);
    }

    @Singleton
    @Provides
    Retrofit provideReprofit(){
        return RetrofitClient.getRetrofitInstance();
    }

    @Singleton
    @Provides
    AppRepository provideAppRepository(Retrofit retrofit) {
        return new AppRepository(retrofit);
    }

}

File: app/src/main/java/beans/prj/newsapp/di/AppModule.kt
package beans.prj.newsapp.di

import android.content.Context
import beans.prj.newsapp.NewsApplication
import beans.prj.newsapp.data.AppRepository
import beans.prj.newsapp.data.RetrofitClient.getRetrofitInstance
import beans.prj.newsapp.ui.base.ViewModelProviderFactory
import dagger.Module
import dagger.Provides
import retrofit2.Retrofit
import javax.inject.Singleton

@Module
class AppModule {
    @Singleton
    @Provides
    fun provideContext(application: NewsApplication): Context {
        return application
    }

    @Singleton
    @Provides
    fun provideViewModelProviderFactory(repository: AppRepository?): ViewModelProviderFactory {
        return ViewModelProviderFactory(repository)
    }

    @Singleton
    @Provides
    fun provideReprofit(context: Context?): Retrofit? {
        return getRetrofitInstance(context!!)
    }

    @Singleton
    @Provides
    fun provideAppRepository(retrofit: Retrofit?): AppRepository {
        return AppRepository(retrofit!!)
    }
}